### ISSUE TITLE
How to structure queue to improve training time

### ISSUE THREAD

**Issue by sarthakpati** [fepegar/torchio#288]
Hi,

I am trying to parameterize [torchio.queue](https://torchio.readthedocs.io/data/patch_training.html?#queue) to improve training time. Here are my augmentations:

```python
global_augs_dict = {
    'normalize':ZNormalization(),
    'affine':RandomAffine(image_interpolation = 'linear'), 
    'elastic': RandomElasticDeformation(num_control_points=(7, 7, 7),locked_borders=2),
    'motion': RandomMotion(degrees=10, translation = 10, num_transforms= 2, image_interpolation = 'linear', p = 1., seed = None), 
    'ghosting': RandomGhosting(num_ghosts = (4, 10), axes = (0, 1, 2), intensity = (0.5, 1), restore = 0.02, p = 1., seed = None),
    'bias': RandomBiasField(coefficients = 0.5, order= 3, p= 1., seed = None), 
    'blur': RandomBlur(std = (0., 4.), p = 1, seed = None), 
    'noise':RandomNoise(mean = 0, std = (0, 0.25), p = 1., seed = None) , 
    'swap':RandomSwap(patch_size = 15, num_iterations = 100, p = 1, seed = None) 
}
```
I have tried a couple of different configurations of the queue construction (for BraTS 2020 training data):

1)

```python
patches_queue = torchio.Queue(
subjects_dataset,max_length = 1,
samples_per_volume  = 1,
sampler = UniformSampler([128,128,128]),
num_workers=4,
shuffle_subjects=False, 
shuffle_patches=True)
```

Here is my output:

<details><summary>Output</summary>
<p>

```bash
Hostname   : XXXXXX
Training Data Samples:  332
CUDA_VISIBLE_DEVICES:  0
Current Device :  0
Device Count on Machine :  1
Device Name :  Tesla P100-PCIE-12GB
Cuda Availibility :  True
Using device: cuda
Memory Usage:
  Allocated: 0.0 GB
  Cached:  0.0 GB
Starting Learning rate is: 0.001


Epoch Started at: 2020-08-20 11:07:31.169938
Epoch # :  0
Learning rate: 0.001
Epoch Training dice: 0.31517219005821734
Best Training Dice: 0.31517219005821734
Average Training Loss: 0.6848278099417825
Best Training Epoch:  0
Epoch Validation dice: 0.4284125193288893
Best Validation Dice: 0.4284125193288893
Average Validation Loss: 0.5715874806711106
Best Validation Epoch:  0
Time for epoch: 730.7090194821358 mins


Epoch Started at: 2020-08-20 23:18:13.753733
Epoch # :  1
Learning rate: 0.00075025
Epoch Training dice: 0.4328200096387767
Best Training Dice: 0.4328200096387767
Average Training Loss: 0.5671799903612232
Best Training Epoch:  1
Epoch Validation dice: 0.4653519422259088
Best Validation Dice: 0.4653519422259088
Average Validation Loss: 0.5346480577740912
Best Validation Epoch:  1
Time for epoch: 727.3753080646197 mins


Epoch Started at: 2020-08-21 11:25:36.334792
Epoch # :  2
Learning rate: 0.0005005
Epoch Training dice: 0.4529779154852848
Best Training Dice: 0.4529779154852848
Average Training Loss: 0.5470220845147151
Best Training Epoch:  2
Epoch Validation dice: 0.49953905589465886
Best Validation Dice: 0.49953905589465886
Average Validation Loss: 0.5004609441053411
Best Validation Epoch:  2
Time for epoch: 738.0272558371227 mins


Epoch Started at: 2020-08-21 23:43:38.034053
Epoch # :  3
Learning rate: 0.0002507499999999999
Epoch Training dice: 0.4745635877166486
Best Training Dice: 0.4745635877166486
Average Training Loss: 0.5254364122833516
Best Training Epoch:  3
Epoch Validation dice: 0.47297200991573
Best Validation Dice: 0.49953905589465886
Average Validation Loss: 0.5270279900842699
Best Validation Epoch:  2
Time for epoch: 728.7016223390897 mins


Epoch Started at: 2020-08-22 11:52:20.179293
Epoch # :  4
Learning rate: 9.999999999999159e-07
Epoch Training dice: 0.4744484525172589
Best Training Dice: 0.4745635877166486
Average Training Loss: 0.5255515474827412
Best Training Epoch:  3
Epoch Validation dice: 0.48653238308167823
Best Validation Dice: 0.49953905589465886
Average Validation Loss: 0.5134676169183217
Best Validation Epoch:  2
Time for epoch: 724.4470616658529 mins


Epoch Started at: 2020-08-22 23:56:47.029944
Epoch # :  5
Learning rate: 0.0002507499999999999
Epoch Training dice: 0.47198090723001274
Best Training Dice: 0.4745635877166486
Average Training Loss: 0.5280190927699873
Best Training Epoch:  3
Epoch Validation dice: 0.45542112143185937
Best Validation Dice: 0.49953905589465886
Average Validation Loss: 0.5445788785681406
Best Validation Epoch:  2
Time for epoch: 723.8550246238708 mins


Epoch Started at: 2020-08-23 12:00:38.396858
Epoch # :  6
Learning rate: 0.0005005
Epoch Training dice: 0.4634351750670278
Best Training Dice: 0.4745635877166486
Average Training Loss: 0.5365648249329721
Best Training Epoch:  3
```

</p>
</details>

2) 

```python
patches_queue = torchio.Queue(
subjects_dataset,max_length = 10,
samples_per_volume  = 10,
sampler = UniformSampler([128,128,128]),
num_workers=4,
shuffle_subjects=False, 
shuffle_patches=True)
```

Here is my output:

<details><summary>Output</summary>
<p>

```bash
Hostname   : XXXXXX
Training Data Samples:  3320
CUDA_VISIBLE_DEVICES:  0
Current Device :  0
Device Count on Machine :  1
Device Name :  Tesla P100-PCIE-12GB
Cuda Availibility :  True
Using device: cuda
Memory Usage:
  Allocated: 0.0 GB
  Cached:  0.0 GB
Starting Learning rate is: 0.001


Epoch Started at: 2020-08-22 09:31:48.666395
Epoch # :  0
Learning rate: 0.001
Epoch Training dice: 0.436339787389386
Best Training Dice: 0.436339787389386
Average Training Loss: 0.5636602126106139
Best Training Epoch:  0
Epoch Validation dice: 0.4510045856401002
Best Validation Dice: 0.4510045856401002
Average Validation Loss: 0.5489954143599001
Best Validation Epoch:  0
Time for epoch: 775.5885859568914 mins


Epoch Started at: 2020-08-22 22:27:23.982092
Epoch # :  1
Learning rate: 0.00075025
Epoch Training dice: 0.47390382915650414
Best Training Dice: 0.47390382915650414
Average Training Loss: 0.5260961708434958
Best Training Epoch:  1
Epoch Validation dice: 0.46360749397146783
Best Validation Dice: 0.46360749397146783
Average Validation Loss: 0.536392506028532
Best Validation Epoch:  1
Time for epoch: 758.6359572728475 mins


Epoch Started at: 2020-08-23 11:06:02.140065
Epoch # :  2
Learning rate: 0.0005005
Epoch Training dice: 0.4902666910009399
Best Training Dice: 0.4902666910009399
Average Training Loss: 0.5097333089990602
Best Training Epoch:  2
Epoch Validation dice: 0.4494907715524214
Best Validation Dice: 0.46360749397146783
Average Validation Loss: 0.5505092284475785
Best Validation Epoch:  1
Time for epoch: 745.3841615160306 mins


Epoch Started at: 2020-08-23 23:31:25.190341
Epoch # :  3
Learning rate: 0.0002507499999999999
Epoch Training dice: 0.501831035780201
Best Training Dice: 0.501831035780201
Average Training Loss: 0.498168964219799
Best Training Epoch:  3
```

</p>
</details>

Is there any way to improve the training time?

Cheers,
Sarthak

**Comment by fepegar** [comment]
Hi again, @sarthakpati.

Very good question. Here are some ideas.

## Measure the run time for each transform

```python
colin = tio.datasets.Colin27()
colin.load()

transformed = colin
for transform in global_augs_dict.values():
    print(transform.name)
    tic = time.time()
    transformed = transform(transformed)
    toc = time.time()
    print(f'{toc - tic:.1f} seconds')
    print()
```

```
ZNormalization
0.1 seconds

RandomAffine
1.3 seconds

RandomElasticDeformation
39.6 seconds

RandomMotion
8.1 seconds

RandomGhosting
4.0 seconds

RandomBiasField
4.3 seconds

RandomBlur
1.1 seconds

RandomNoise
0.1 seconds

RandomSwap
0.1 seconds
```

## Use `OneOf`

It's probably not necessary to apply a random affine and a random elastic deformation to each image. You could use `OneOf` to apply the affine to most images, which is much faster*.

The same applies to the k-space artifacts.

```python
spatial_transform = OneOf({
    RandomAffine(): 0.8,
    RandomElasticDeformation(): 0.2,
})

mri_artifact = OneOf({
    RandomMotion(): 0.5,
    RandomGhosting(): 0.5,
})

global_augs_dict = {
    'normalize': ZNormalization(),
    'spatial': spatial_transform,
    'kspace': mri_artifact,
    'bias': RandomBiasField(coefficients = 0.5, order= 3, p= 1., seed = None), 
    'blur': RandomBlur(std = (0., 4.), p = 1, seed = None), 
    'noise': RandomNoise(mean = 0, std = (0, 0.25), p = 1., seed = None) , 
    'swap': RandomSwap(patch_size = 15, num_iterations = 100, p = 1, seed = None) 
}
```

*I just realised there's a bug in `RandomElasticDeformation`: it should use nearest neighbor interpolation by default for label maps, but it's not. I'll fix that once I'm done with this comment.

## Use the probability kwarg `p`

You might not want to apply the k-space artifact transform to all images, as it a but slow. Same for the bias field. Maybe you can just apply it to every other image:

```python
mri_artifact = OneOf({
    RandomMotion(): 0.5,
    RandomGhosting(): 0.5,
}, p=0.5)
```

## Use a longer queue

You are using a queue of length 1. Assuming each subject has a T1, T2, FLAIR and segmentation, that's 128**3 * 4 * 4 = 32 MiB per patch. Your queue could easily store 100 patches instead, so I recommend you use a much larger `max_length`. For example, 100. It depends on how much RAM you have available.

## Sample more patches per volume

If you sample only 1 patch from a volume, you need to load and transform the whole volume for just one patch. It's much more efficient to extract multiple patches each time the subject is loaded & transformed. For example, 10 if your patch is large (128), or 30 if it's smaller (64).

---

Other considerations:
1. Why do you use `RandomSwap` here? What is the task you're trying to solve?
2. Is there a reason why you're not shuffling your subjects? It's typically beneficial to shuffle the training set to improve stochasticity during the optimization process

Does that make sense?

**Comment by fepegar** [comment]
Apparently, other people are complaining about how slow the elastic deformation is:

https://discourse.itk.org/t/why-resampleimagefilter-is-slow/1217/12

**Comment by sarthakpati** [comment]
Thanks a ton! I will try some of these out and let you know.

**Comment by fepegar** [comment]
Thanks. Please do let me know if you get better times using my suggestions.

**Comment by sarthakpati** [comment]
Some updates to the questions/comments you posed:

- I added the `OneOff` and started the training (will update once I start getting some results). I won't be using the probability arg for now (will see if these changes help and then will look into it)
- Using `max_length = 100` and `samples_per_volume = 10` (I have patch size of `128**3`).
- Unfortunately, cannot use a smaller image as I am trying to train a generalizable model; if I was looking to participate in BraTS, I would definitely have cropped the image(s) to fit the atlas space.
- AMP is something I will be adding to my framework soon (thanks for reminding me!). 
- `RandomSwap` is an option I am giving to the user, they can use/disregard as they see fit; hoping that this makes the model more generalizable
- Dataset shuffle is getting handled in a separate layer, so `torchio` always gets a shuffled set.

Thanks, again!

**Comment by sarthakpati** [comment]
With the following parameters:

```python
spatial_transform = OneOf({
    RandomAffine(): 0.8,
    RandomElasticDeformation(): 0.2,
})

mri_artifact = OneOf({
    RandomMotion(): 0.5,
    RandomGhosting(): 0.5,
})

augmentations = {
    'normalize':ZNormalization(),
    'spatial': spatial_transform,
    'noise':RandomNoise(mean = 0, std = (0, 0.25), p = 1., seed = None) , 
    'swap':RandomSwap(patch_size = 15, num_iterations = 100, p = 1, seed = None) 
}

max_length = 30
samples_per_volume = 10
patch_size = [128,128,128] # brats dataset
num_workers = 4
batch_size = 1
RAM = 40G # the training session exceeds this
```

The run-time per epoch is about 177 minutes. Is this expected?

**Comment by fepegar** [comment]
> The run-time per epoch is about 177 minutes. Is this expected?

I'm glad it's now more than 4x faster. It's hard to say whether that's expected. It depends on many things: your processor speed, the size of your dataset, the size of your network, the size of your images... You could decrease it even further by using 0.9 and 0.1 for the spatial transforms probabilities. Using AMP would also help, as I said.

I'm not that familiar with the BraTS dataset, but cropping the images beforehand as much as you can would also help. I'm not sure what you mean by "cannot use a smaller image as I am trying to train a generalizable model". We all try to train generalizable models!

**Comment by sarthakpati** [comment]
> I'm glad it's now more than 4x faster

The previous runtime for this augmentation list for **30 minutes**.

**Comment by fepegar** [comment]
You hadn't mentioned that. What do you mean by previous? Which version of TorchIO were you using previously?

**Comment by sarthakpati** [comment]
By "previous", I mean my original post.

I have been using 0.17.21

**Comment by fepegar** [comment]
Sorry, I don't understand. Do you mean your first comment on this issue (https://github.com/fepegar/torchio/issues/288#issue-684724423)?

The logs you shared read `Time for epoch: 730.7090194821358 mins`.

**Comment by sarthakpati** [comment]
Yeah, that is right. 

> The logs you shared read Time for epoch: 730.7090194821358 mins. 

This time is with this augmentation list:

```python
global_augs_dict = {
    'normalize':ZNormalization(),
    'affine':RandomAffine(image_interpolation = 'linear'), 
    'elastic': RandomElasticDeformation(num_control_points=(7, 7, 7),locked_borders=2),
    'motion': RandomMotion(degrees=10, translation = 10, num_transforms= 2, image_interpolation = 'linear', p = 1., seed = None), 
    'ghosting': RandomGhosting(num_ghosts = (4, 10), axes = (0, 1, 2), intensity = (0.5, 1), restore = 0.02, p = 1., seed = None),
    'bias': RandomBiasField(coefficients = 0.5, order= 3, p= 1., seed = None), 
    'blur': RandomBlur(std = (0., 4.), p = 1, seed = None), 
    'noise':RandomNoise(mean = 0, std = (0, 0.25), p = 1., seed = None) , 
    'swap':RandomSwap(patch_size = 15, num_iterations = 100, p = 1, seed = None) 
}
```

**Comment by sarthakpati** [comment]
After adding AMP, with augmentations:

```python
augmentations = {
    'normalize':ZNormalization(),
    'noise':RandomNoise(mean = 0, std = (0, 0.25), p = 1., seed = None) , 
    'swap':RandomSwap(patch_size = 15, num_iterations = 100, p = 1, seed = None) 
}
```
And queue properties:
```
max_length = 20
samples_per_volume = 10
patch_size = [128,128,128] # brats dataset
num_workers = 4
batch_size = 1
```
The training per epoch has gone down from ~150 minutes to ~100 minutes. 

This **still** a lot and I think I am missing some understanding regarding how to use the queue properly, here...

**Comment by fepegar** [comment]
If you change the mixed precision settings and the transforms list and the queue parameters at the same time, it's hard to know what causes what.

I suggest you change one setting at the time to better understand the influence on the training time.

Using a larger batch size and number of workers would also help, but I suppose you have 4 cores and the GPU can't hold more than one patch at the time, right?

**Comment by sarthakpati** [comment]
> If you change the mixed precision settings and the transforms list and the queue parameters at the same time, it's hard to know what causes what. 

I agree, the list that I posted in my first entry is still training with AMP, so I figured I'd give the results for the least number of augmentations first.

> Using a larger batch size and number of workers would also help, but I suppose you have 4 cores and the GPU can't hold more than one patch at the time, right?

The number of CPU cores won't be an issue for me and I can easily scale that. I am running these training jobs on nodes with P100s, but yeah, it cannot hold more than 1 patch at a time.

**Comment by meghbhalerao** [comment]
> transformed = colin
> for transform in global_augs_dict.values():
>     print(transform.name)
>     tic = time.time()
>     transformed = transform(transformed)
>     toc = time.time()
>     print(f'{toc - tic:.1f} seconds')
>     print()
> ```
> 
> ```
> ZNormalization
> 0.1 seconds
> 
> RandomAffine
> 1.3 seconds
> 
> RandomElasticDeformation
> 39.6 seconds
> 
> RandomMotion
> 8.1 seconds
> 
> RandomGhosting
> 4.0 seconds
> 
> RandomBiasField
> 4.3 seconds
> 
> RandomBlur
> 1.1 seconds
> 
> RandomNoise
> 0.1 seconds
> 
> RandomSwap
> 0.1 seconds
> ```
Hi @fepegar,
Are these times approximately similar  across different datasets? Or is there a drastic change depending on voxel resolution, image size etc.? Any idea?
Thanks,
Megh

**Comment by romainVala** [comment]
Hello
it is dependent on images size, the timing you show are similar (at least the magnitude order) to what I have with 182x218x182 volumes.
may be RandomElasticDeformation is quite big, (I have to check ...), but RandomMotion and BiasField are similar to what I remember
what is your volume size ?

**Comment by meghbhalerao** [comment]
My volume size is `144 * 144 * 32`, and each patient has **3 modalities**. I am doing the patch based training, with 1 patch per image, bringing the total number of patches to 104. Yes, the `RandomElasticDeformation`, is quite big.

The augmentation I am using is:
```
spatial_transform = OneOf({
    RandomAffine(): 0.2,
    RandomElasticDeformation(): 0.1,
})
```
My training time per epoch for a **3D UNet** is approx **140 mins**. Is this expected?

**Comment by fepegar** [comment]
@romainVala 

The times I shared are from colin 1998, so 3 images of size 181 x 217 x 181.

```python
In [1]: import torchio

In [2]: colin = torchio.datasets.Colin27()
Using cache found in /Users/fernando/.cache/torchio/mni_colin27_1998_nifti

In [3]: colin.spatial_shape
Out[3]: (181, 217, 181)

In [4]: colin
Out[4]: Colin27(Keys: ('t1', 'head', 'brain'); images: 3)
```

---

@meghbhalerao 

The [docs for `OneOf`](https://torchio.readthedocs.io/transforms/augmentation.html#oneof) say "Probabilities are normalized so they sum to one". Therefore, in your transform, the elastic deformation is being applied with a probability of `0.33` and the affine, with `0.66`.

> My training time per epoch for a 3D UNet is approx 140 mins. Is this expected?

As I said above, there are many factors that contribute to an epoch duration. You can try the different changes I suggested here to learn about the influence each of them has on the run time.

**Comment by meghbhalerao** [comment]
Hi @fepegar,
Thank you. I timed different lines of codes in my script, and I figured the bottleneck is from the augmentations (the step which fetches each batch i.e. `batch = next(train_loader)`). However, I will try further combinations to see if I am able to improve the per epoch time. Thanks again for your help.
Megh

**Comment by meghbhalerao** [comment]
> ```python
> mri_artifact = OneOf({
>     RandomMotion(): 0.5,
>     RandomGhosting(): 0.5,
> }, p=0.5)
> ```
Hi guys,
When I try to do `mri_aritifact(p=some_val)`, I get the following error:
```python
TypeError: __call__() got an unexpected keyword argument 'p'
```
Can anyone help?
Thanks

**Comment by sarthakpati** [comment]
Write it as a function:

```python
def mri_artifact(prob = 1):
    return OneOf({RandomMotion(): 0.5, RandomGhosting(): 0.5,}, p=prob)
```

And use it as:

```python
mri_artifact(prob=1)
```

**Comment by fepegar** [comment]
@meghbhalerao 

> Hi @fepegar,
Thank you. I timed different lines of codes in my script, and I figured the bottleneck is from the augmentations (the step which fetches each batch i.e. batch = next(train_loader)). However, I will try further combinations to see if I am able to improve the per epoch time. Thanks again for your help.
Megh

1. You don't need to use `next` to extract a batch from a data loader. You can just iterate over it: `for batch in train_loader`.
2. I'm surprised the bottleneck is in that specific line. Are you sure that's the case? When one uses a queue, there are two data loaders working. The first is internal to the queue and spawns `num_workers` processes to load and process the volumes from the dataset. This is typically the bottleneck. The second one is the one you instantiate to collate batches from the queue. This is a cheap operation as it just pops batches from a Python list.

> When I try to do mri_aritifact(p=some_val), I get the following error:

When you define a `__call__` method in a `class`, that's what's called when you 'call' the class:

```python
In [11]: class Callable:
    ...:     def __init__(self):
    ...:         print('Instantiating...')
    ...:     def __call__(self):
    ...:         print('Calling...')
    ...:

In [12]: transform = Callable()
Instantiating...

In [13]: transform()
Calling...
```

`p` is a kwarg of the `__init__` method, not of the `__call__` method. Does that make sense?

More info: https://stackoverflow.com/questions/9663562/what-is-the-difference-between-init-and-call

**Comment by dvolgyes** [comment]
I haven't finished writing a fix yet, but one of my issues with Queue that fills the list first, randomizes the list, and then yields items.
For me, filling the queue takes extremely long time, but i have many cores, so it would be more helpful for me 
a queue which yields the first available sample, and doesn't wait for the queue to be filled. 
When i have a bit time, i will submit an alternative implementation.
(Basically: randomizing all samples in advance, putting them into a task queue where they wait, and having a processing queue with e.g. 30 spots. Whenever the processing queue has a ready item, yield it, whenever it has an empty spot, start a new task in parallel in the background. It is not the same as the current behavior, but according to my own tests, majority of my problems comes from the fact that the queue must be filled first before it yields any sample.) 

Long story short: i would measure/profile how much time it takes to fill the queue, read files, and other tasks, not just the length of the augmentation.

**Comment by fepegar** [comment]
Having that implementation would be great, @dvolgyes! I've always thought it would be a much better option, but I don't know much about parallel processing, so I didn't know where to start.

Let me know if I can help.

**Comment by sarthakpati** [comment]
I think another mechanism to improve run-time could be using SimpleITK for I/O instead of Nibabel. SITK uses C++ backend, which is far more efficient compared to Nibabel's Python. This would also make the transforms and API much easier to manage (no need to save separate affine structure, can simply pass the SITK image around, which contains everything).

**Comment by dvolgyes** [comment]
Let's move conversation to a new queue implementation thread: #301

**Comment by fepegar** [comment]
@sarthakpati, can you upgrade the library to `v0.17.36` and try again? `UniformSampler` is much faster after #296.

**Comment by dvolgyes** [comment]
One remark: using UniformSampler makes patch generation very fast. So your code will be faster if you use less number of volumes and more patches per volume. If your volume is reasonable large compared to the patch, you can easily get e.g. 32 or 64 different patches, without being to self-similar. But of course, the optimal parameters depend on the data.

**Comment by sarthakpati** [comment]
Thanks, @fepegar and @dvolgyes. I will try the updated version and let you know. I will also keep a track of #301. Cheers!

**Comment by fepegar** [comment]
Closing due to inactivity.
